prologin/workernode/operations.py

```diff
@@ -46,308 +50,319 @@ def untar(content, path, compression='gz'):
-        config['path']['stechec_client'],
-        "--name", str(pl_id),
-        "--rules", config['path']['rules'],
-        "--champion", champion_path + '/champion.so',
-        "--req_addr", req_addr,
-        "--sub_addr", sub_addr,
-        "--socket_timeout", "45000",
-        "--time", "1500",
-        "--verbose", "1",
-    ]
-    # fmt: on
-    cmd += ["--client_id", str(order_id)] if order_id is not None else []
-
-    if map_contents is not None:
-        f = tempfile.NamedTemporaryFile(mode='w')
-        f.write(map_contents)
-        f.flush()
-        os.chmod(f.name, 0o644)
-        cmd += ['--map', f.name]
-
-    # Build resource limits
-    limits = {
-        'wall-time': config['isolate'].get('time_limit_secs', 350),
-        'mem': config['isolate'].get('mem_limit_MiB', 500) * 1000,
-        'processes': config['isolate'].get('processes', 50),
-        'fsize': 256,
-    }
-    allowed_dirs = [
-        '/var',
-        '/etc',
-        '/tmp',
-        sockets_dir + ':rw',
-        champion_path,
-        os.path.dirname(config['path']['stechec_client']),
-        os.path.dirname(config['path']['rules']),
-    ]
-
-    # Remove memory limit if Java
-    if os.path.exists(os.path.join(champion_path, 'Prologin.class')):
-        limits.pop('mem')
-
-    # Run the isolated client
-    result = await isolate_communicate(
-        cmd, limits, env=env, allowed_dirs=allowed_dirs, merge_outputs=True
-    )
-    return result.isolate_retcode, get_output(result)
-
 
-# TODO: refactor return value to a dataclass
-async def spawn_match(config, match_id, players, map_contents):
+class SpawnClient(Operation):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.result = {
+            'champion_id': None,
+            **self.result,
+        }
+        self.limits = {
+            'wall-time': self.config['isolate'].get('time_limit_secs', 350),
+            'mem': self.config['isolate'].get('mem_limit_MiB', 500) * 1000,
+            'processes': self.config['isolate'].get('processes', 50),
+            'fsize': 256,
+        }
+        self.allowed_dirs = [
+            '/var',
+            '/etc',
+            '/tmp',
+            os.path.dirname(self.config['path']['stechec_client']),
+            os.path.dirname(self.config['path']['rules']),
+        ]
+
+    def extract_champion(self, champion_tgz_b64):
+        ctgz = b64decode(champion_tgz_b64)
+        untar(ctgz, self.isolator.path)
+        return Path('/box')
+
+    async def _run(
+        self,
+        *,
+        sockets_dir,
+        req_addr,
+        sub_addr,
+        champion_id,
+        player_id,
+        champion_tgz_b64,
+        map_content=None,
+        order_id=None,
+    ):
+        self.result['champion_id'] = champion_id
+
+        self.allowed_dirs.append(sockets_dir + ':rw')
+
+        champion_path = self.extract_champion(champion_tgz_b64)
+        env = {'CHAMPION_PATH': champion_path, 'HOME': '/tmp'}
```

Comment:
I'd explicitly call `str()` on the `Path` just in case. You don't want random implicit `repr()` inside things that use `env` to break this. This applies to other places too I believe.